rclpy/rclpy/executors.py

```diff
@@ -769,7 +770,7 @@ def handler_wrapper(handler):
                 handler()
                 if handler.exception() is not None:
                     raise handler.exception()
-            self._executor.submit(handler_wrapper(handler))
+            self._executor.submit(partial(handler_wrapper, handler))
```

Comment:
This also works the same.

```suggestion
            self._executor.submit(handler_wrapper, handler)
```

Reply:
Well, I didn't notice it in my original https://github.com/ros2/rclpy/pull/984 review, but that PR is broken.
It's not submiting `handler` for execution in the thread pool.
It's execution the handler immediately and submitting `None` to the executor ...

The only way to detect exceptions is to store the future returned by the executor and check if they finished successfully or with an exception.
We cannot block until the future is completed, so in each iteration of `spin()` we could poll all the stored futures and check if one of them failed (removing the ones that completed).
When a failure is detecter, it's thrown.

---

I'm okay with reverting https://github.com/ros2/rclpy/pull/984 temporally, as the solution to the actual problem #984 tried to solve isn't that trivial.

Reply:
@wjwwood what do you think?

Reply:
> I'm okay with reverting https://github.com/ros2/rclpy/pull/984 temporally, as the solution to the actual problem https://github.com/ros2/rclpy/pull/984 tried to solve isn't that trivial.

agree.

Reply:
BTW, thanks for the PR @fujitatomoya !
I didn't undestand what was the issue until this :sweat_smile:

Reply:
Reopen revert commit https://github.com/ros2/rclpy/pull/1017